Philips.CodeAnalysis.MaintainabilityAnalyzers/RuntimeFailure/ThrowInnerExceptionAnalyzer.cs

```diff
@@ -0,0 +1,97 @@
+﻿// © 2020 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.RuntimeFailure
+{
+	/// <summary>
+	/// Include the original exception when rethrowing.
+	/// </summary>
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class ThrowInnerExceptionAnalyzer : DiagnosticAnalyzer
+	{
+		private const string Title = "User inner exceptions for unhandled exceptions";
+		private const string Message = "Rethrown exception should include caught exception.";
+		private const string Description = "User inner exceptions for unhandled exceptions";
+		private const string Category = Categories.RuntimeFailure;
+
+		private static readonly DiagnosticDescriptor Rule =
+			new DiagnosticDescriptor(
+				Helper.ToDiagnosticId(DiagnosticIds.ThrowInnerException),
+				Title,
+				Message,
+				Category,
+				DiagnosticSeverity.Error,
+				isEnabledByDefault: true,
+				description: Description
+			);
+
+		/// <summary>
+		/// <inheritdoc/>
+		/// </summary>
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		/// <summary>
+		/// <inheritdoc/>
+		/// </summary>
+		public override void Initialize(AnalysisContext context)
+		{
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.EnableConcurrentExecution();
+			context.RegisterSyntaxNodeAction(
+				AnalyzeCatchException,
+				SyntaxKind.CatchClause
+			);
+		}
+
+		private void AnalyzeCatchException(SyntaxNodeAnalysisContext context)
+		{
+			var catchNode = (CatchClauseSyntax)context.Node;
+			// Look for throw statements and check them.
+			var hasBadThrowNodes = catchNode.DescendantNodes().OfType<ThrowStatementSyntax>()
+				.Where(node => !IsCorrectThrow(context, node)).Any();
+			if (hasBadThrowNodes)
+			{
+				var location = catchNode.GetLocation();
```

Comment:
using the location of the catchNode makes the experience in Visual Studio not great.  I think you should put it on the individual throwStatements (and accept that if you have 2 different throws in a single catch block you need 2 errors)

Reply:
Done